docs/rewriter/examples/broadcast_matmul.py

```diff
@@ -0,0 +1,216 @@
+"""Onnx Pattern Rewriting with match condition parameter.
+
+This script shows how to define a rewriting rule based on patterns while
+utilizing the match condition parameter.
+
+First we write a dummy model with a several Reshape nodes and a Matmul node
+===================
+"""
+
+import logging
+
+import numpy as np
+
+import onnx
+
+import onnxscript
+from onnxscript import FLOAT, opset18, script
+from onnxscript.rewriter import _ir_utils, pattern
+
+logger = logging.getLogger(__name__)
+
+
+@script()
+def original_model(A: FLOAT[1, 4, 512, 512], B: FLOAT[1, 4, 512, 64]) -> FLOAT[1, 4, 512, 64]:
+    # NOTE: Modified from `value_ints` to `value`
+    shape_a = opset18.Constant(value=[4, 512, 512])
+    reshape_a = opset18.Reshape(A, shape_a)
+    shape_b = opset18.Constant(value=[4, 512, 64])
+    reshape_b = opset18.Reshape(B, shape_b)
+    matmul = opset18.MatMul(reshape_a, reshape_b)
+    shape_c = opset18.Constant(value=[1, 4, 512, 64])
+    result = opset18.Reshape(matmul, shape_c)
+    return result
+
+
+model = original_model.to_model_proto()
+onnx.checker.check_model(model)
+
+
+####################################
+# The target pattern
+# =====================
+
+op = pattern.onnxop
+
+
+def two_reshapes_matmul_reshape_pattern(input_a, input_b, shape_a, shape_b, shape_c):
+    reshape_a = op.Reshape(input_a, shape_a)
+    reshape_b = op.Reshape(input_b, shape_b)
+    matmul = op.MatMul(reshape_a, reshape_b)
+    return op.Reshape(matmul, shape_c)
+
+
+####################################
+# The replacement pattern
+# =====================
+
+
+def matmul(op, input_a, input_b, **_):
+    return op.MatMul(input_a, input_b)
+
+
+####################################
+# Write condition to check if we need to replace the pattern
+# =====================
+
+
+def check_if_need_reshape(match_bindings) -> bool:
```

Comment:
We should migrate to a simpler interface here. Ideally, the caller should pass in `match_bindings` as `**match_bindings`. We can then just define this as a function with inputs like `(input_a, input_b, shape_c, **_)` and omit all references to `match_bindings` down below. This requires a couple of lines of code change as well. Mentioning it here, whether it is done as part of this PR or as another follow-up

Reply:
That's a great point. Modified how `match_bindings` are processed are part of this PR. @titaiwangms Feel free to take a look at changes related to match_condition functions.

Reply:
I will take a look today.

Reply:
I think this is the same suggestion that Rama gave me when I wrote https://github.com/microsoft/onnxscript/blob/53e6832fdc7d8f3f2cf2e2f2acf397073b87f1aa/onnxscript/rewriter/onnxruntime/instance_to_group_normalization.py#L19-L28. To make it more explicit, we should write those values out of match_bindings. This looks nice!

Reply:
Thanks!

Reply:
That's a good idea. Added